include/openenclave/host.h

```diff
@@ -64,6 +65,60 @@ typedef void (*oe_ocall_func_t)(
     size_t output_buffer_size,
     size_t* output_bytes_written);
 
+/**
+ * The following structures are used by context-switchless calls, which is
+ * experimental, and subject to changes.
+ */
+#ifdef OE_CONTEXT_SWITCHLESS_EXPERIMENTAL_FEATURE
+
+/**
+ * Types of configurations passed into **oe_create_enclave**
+ */
+typedef enum _oe_enclave_config_type
+{
+    OE_ENCLAVE_CONFIG_CONTEXT_SWITCHLESS = 0xdc73a628,
+} oe_enclave_config_type_t;
+
+/**
+ * The configuration for context-switchless calls.
+ */
+typedef struct _oe_enclave_config_context_switchless
+{
+    /**
+     * The max number of worker threads for context-switchless ocalls.
+     * The actual number of threads launched could be capped for performance
+     * reasons.
+     */
+    size_t max_host_workers;
+    /**
+     * Context-switchless ecalls are not enabled yet. The max number of enclave
+     * workers should be 0.
+     */
+    size_t max_enclave_workers;
+} oe_enclave_config_context_switchless_t;
+
+/**
+ * The uniform structure type containing a specific type of enclave
+ * configuration.
+ */
+typedef struct _oe_enclave_config
+{
+    /**
+     * The type of the configuration in **u**
+     */
+    oe_enclave_config_type_t config_type;
+    /**
+     * The specific configuration for the enclave, such as configuring
+     * context-switchless calls.
+     */
+    union {
+        const oe_enclave_config_context_switchless_t* context_switchless_config;
```

Comment:
What is the rationale for a union of struct pointers instead of a union of structs?
My recollection of the discussions was that u would be a union of config structures.

@mikbras @dthaler

Reply:
If `u`  is a union of config structures, the struct `oe_enclave_config_t` doesn't have a fixed size, and we won't fit different types of configs into an array.

Reply:
If it is a union of config structures, the struct will still have a fixed size. Additionally one would be able to initialize the struct in place.

Reply:
We will add more configuration types. Not all of them will have the same length as `oe_enclave_config_context_switchless_t`. Using pointers, not the structure itself, in the union, ensures the union size is fixed, hence the size of `oe_enclave_config_t` is fixed .

Reply:
The size of union will always be the size of its largest member. You are right that `oe_enclave_config_t` is size fixed even if we put config structures, instead of pointers, in it. Now I am not sure what was the reason we decided to put pointers in the union.

@dthaler @mikbras

Reply:
The only reason to use pointers is just memory. A union is the size of its largest member.  So if you have a small member, and a large member, then an large array of small members will be proportionately larger.  With pointers, you waste the size of a pointer per array element, but that's all.

Reply:
Sounds good.